### ISSUE TITLE
Function svds freezes

### ISSUE THREAD

**Issue by lostella** [JuliaLang/julia#20303]
I'm experiencing a problem with `svds` in Julia 0.6.0-dev.2439, commit 953029f, on macOS. What happens is that the REPL freezes after invoking svds, as in

```
x = randn(10,10)
out = svds(x, nsv=3)
```

After cmd+c'ing and re-issuing the latter command, the output is produced correctly.

``versioninfo()`` gives:

```
Julia Version 0.6.0-dev.2439
Commit 953029f (2017-01-28 07:23 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin13.4.0)
  CPU: Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)
```

**Comment by andreasnoack** [comment]
Confirmed. Inference appears to hang in an infinite loop. Probably caused by #18457.

**Comment by RalphAS** [comment]
On Linux x86_64 it's not an infinite loop - it just takes 100 seconds or so (compared to about 3 seconds in 0.5.0). Profiling points to deeply nested inference code, with the bulk of the time in `jl_matching_methods()` and underlying C code.

Interestingly, there is no such delay for `eigs` with dense or sparse array arguments, just for the type used by `svds`.

**Comment by tkelman** [comment]
It looks like this may have been fixed by https://github.com/JuliaLang/julia/pull/20354 ?

**Comment by simonbyrne** [comment]
It seems to be a bit better: the first call to `svds` now takes 7 seconds on my machine

**Comment by andreasnoack** [comment]
Fixed by 93a725d51dde9bc5731e08fd50085108e3d43f3c
